x/examples/smart-proxy/main.go

```diff
@@ -0,0 +1,147 @@
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"net/http"
+	"os"
+	"os/signal"
+	"strings"
+	"time"
+
+	"github.com/Jigsaw-Code/outline-sdk/transport"
+	"github.com/Jigsaw-Code/outline-sdk/x/config"
+	"github.com/Jigsaw-Code/outline-sdk/x/httpproxy"
+	"github.com/Jigsaw-Code/outline-sdk/x/smart"
+)
+
+var debugLog log.Logger = *log.New(io.Discard, "", 0)
+
+type stringArrayFlagValue []string
+
+func (v *stringArrayFlagValue) String() string {
+	return fmt.Sprint(*v)
+}
+
+func (v *stringArrayFlagValue) Set(value string) error {
+	*v = append(*v, value)
+	return nil
+}
+
+func main() {
+	verboseFlag := flag.Bool("v", false, "Enable debug output")
+	addrFlag := flag.String("localAddr", "localhost:1080", "Local proxy address")
+	configFlag := flag.String("config", "config.json", "Address of the config file")
+	transportFlag := flag.String("transport", "", "The base transport for the connections")
+	var domainsFlag stringArrayFlagValue
+	flag.Var(&domainsFlag, "domain", "The test domains to find strategies.")
+
+	flag.Parse()
+	if *verboseFlag {
+		debugLog = *log.New(os.Stderr, "", log.LstdFlags|log.Lmicroseconds|log.Lshortfile)
+	}
+
+	if len(domainsFlag) == 0 {
+		log.Fatal("Must specify flag --domain")
+	}
+
+	if *configFlag == "" {
+		log.Fatal("Must specify flag --config")
+	}
+
+	finderConfig, err := os.ReadFile(*configFlag)
+	if err != nil {
+		log.Fatalf("Could not read config: %v", err)
+	}
+
+	packetDialer, err := config.NewPacketDialer(*transportFlag)
+	if err != nil {
+		log.Fatalf("Could not create packet dialer: %v", err)
+	}
+	streamDialer, err := config.NewStreamDialer(*transportFlag)
+	if err != nil {
+		log.Fatalf("Could not create stream dialer: %v", err)
+	}
+	if strings.HasPrefix(*transportFlag, "ss:") {
+		innerDialer := streamDialer
+		// Hack to disable IPv6 with Shadowsocks, since it doesn't communicate connection success.
+		streamDialer = transport.FuncStreamDialer(func(ctx context.Context, addr string) (transport.StreamConn, error) {
+			host, _, err := net.SplitHostPort(addr)
+			if err != nil {
+				return nil, err
+			}
+			if ip := net.ParseIP(host); ip != nil && ip.To4() == nil {
+				return nil, fmt.Errorf("IPv6 not supported")
+			}
+			return innerDialer.DialStream(ctx, addr)
+		})
+	}
+	finder := smart.StrategyFinder{
+		LogWriter:    debugLog.Writer(),
+		TestTimeout:  5 * time.Second,
+		StreamDialer: streamDialer,
+		PacketDialer: packetDialer,
+	}
+
+	fmt.Println("Finding strategy")
+	startTime := time.Now()
+	dialer, err := finder.NewDialer(context.Background(), domainsFlag, finderConfig)
```

Comment:
Shall we rename it to `finder.FindWorkingDialer(...)`? `NewDialer` sounds like an instant object creation method, but it actually takes some time to find a dialer.

Reply:
I had that before, but to me "find" sounds like we are picking from an existing set. We are finding a strategy, then creating a dialer. I don't think New needs to be instant. The "finder" object also conveys the finding idea.